src/com/owncloud/android/files/services/FileUploadService.java

```diff
@@ -0,0 +1,1395 @@
+                UploadDbObject previous = mPendingUploads.putIfAbsent(uploadKey, uploadObject);
+                
+                if(previous == null)
+                {
+                    Log_OC.d(TAG, "mPendingUploads added: " + uploadObject.toFormattedString());
+
+                    // if upload was not queued before, we can add it to
+                    // database because the user wants the file to be uploaded.
+                    // however, it can happened that the user uploaded the same
+                    // file before in which case there is an old db entry.
+                    // delete that to be sure we have the latest one.
+                    if(mDb.removeUpload(uploadObject.getLocalPath())>0) {
+                        Log_OC.w(TAG, "There was an old DB entry " + uploadObject.getLocalPath()
+                                + " which had to be removed in order to add new one.");
+                    }
+                    boolean success = mDb.storeUpload(uploadObject);
+                    if(!success) {
+                        Log_OC.e(TAG, "Could not add upload " + uploadObject.getLocalPath()
+                                + " to database. This should not happen.");
+                    } 
+                } else {
+                    Log_OC.w(TAG, "FileUploadService got upload intent for file which is already queued: "
+                            + uploadObject.getRemotePath());
+                    // upload already pending. ignore.
+                }
+            }
+        }
+        // at this point mPendingUploads is filled.
+
+
+
+        Iterator<String> it;
+        if (intent != null && intent.getStringExtra(KEY_RETRY_REMOTE_PATH) != null) {
+            ArrayList<String> list = new ArrayList<String>(1);
+            String remotePath = intent.getStringExtra(KEY_RETRY_REMOTE_PATH);
+            list.add(remotePath);
+            it = list.iterator();
+            // update db status for upload
+            UploadDbObject uploadDbObject = mPendingUploads.get(remotePath);
+            uploadDbObject.setUploadStatus(UploadStatus.UPLOAD_LATER);
+            uploadDbObject.setLastResult(null);
+            mDb.updateUploadStatus(uploadDbObject);
+
+            Log_OC.d(TAG, "Start uploading " + remotePath);
+        } else {
+            it = mPendingUploads.keySet().iterator();
+        }
+        if (it.hasNext()) {
+            while (it.hasNext()) {
+                String upload = it.next();
+                UploadDbObject uploadDbObject = mPendingUploads.get(upload);
+
+                if (uploadDbObject == null) {
+                    Log_OC.e(TAG, "Cannot upload null. Fix that!");
+                    continue;
+                }
+
+                UploadTask uploadTask = new UploadTask(uploadDbObject);
+                uploadExecutor.submit(uploadTask);
+            }
+            StopSelfTask stopSelfTask = new StopSelfTask(intentStartId);
+            uploadExecutor.submit(stopSelfTask);
```

Comment:
Thanks @LukeOwncloud  pull request is the most valuable since the existence of auto upload! 
I found some problems about this pull request. Issue #1129 explain one part of the problem. Files get uploaded multiple times. 
## Scenario:
 1. Start an upload
 2. Start another upload while the other upload is running
this can be done e.g. using auto picture upload and take multiple pictures in sequence. 
## Expected:
every image is uploaded once
## Actual
some images are uploaded multiple times 

The Problem is `onHandleIntent` will be called multiple times. If calling `onHandleIntent` while an upload is running all entries in `mPendingUploads` will be added to the `uploadExecutor`-queue even if they are already in this queue.